cinema/views.py

```diff
@@ -45,8 +70,43 @@ def get_serializer_class(self):
 
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
-    queryset = MovieSession.objects.all()
-    serializer_class = MovieSessionSerializer
+
+    def get_queryset(self):
+        queryset = MovieSession.objects.all()
+
+        if self.action == "list":
+            queryset = (
+                queryset.select_related("movie", "cinema_hall")
+                .annotate(
+                    tickets_available=(
+                        F("cinema_hall__rows")
+                        * F("cinema_hall__seats_in_row")
+                        - Count("tickets")
+                    )
+                )
+            )
+
+        if self.action == "retrieve":
+            queryset = (
+                queryset.select_related("cinema_hall", "movie")
+                .prefetch_related("tickets")
+            )
+
+        date = self.request.query_params.get("date")
+        movie = self.request.query_params.get("movie")
+
+        if date:
+            year, month, day = date.split("-")
+            queryset = queryset.filter(
+                show_time__year=year,
+                show_time__month=month,
+                show_time__day=day
+            )
+
+        if movie:
+            queryset = queryset.filter(movie_id=int(movie))
```

Comment:
Not sure about converting the movie parameter using the int() function. What if not proper type passing - error can occur during converting?